packages/xrpc-server/src/server.ts

```diff
@@ -432,76 +403,35 @@ export class Server {
     }
   }
 
-  private setupRouteRateLimits(nsid: string, config: XRPCHandlerConfig) {
-    this.routeRateLimiters[nsid] = []
-    for (const limit of this.globalRateLimiters) {
-      this.routeRateLimiters[nsid].push({
-        consume: (ctx: XRPCReqContext) => limit.consume(ctx),
-        reset: (ctx: XRPCReqContext) => limit.reset(ctx),
-      })
-    }
-
-    if (config.rateLimit) {
-      const limits = Array.isArray(config.rateLimit)
-        ? config.rateLimit
-        : [config.rateLimit]
-      this.routeRateLimiters[nsid] = []
-      for (let i = 0; i < limits.length; i++) {
-        const limit = limits[i]
-        const { calcKey, calcPoints } = limit
-        if (isShared(limit)) {
-          const rateLimiter = this.sharedRateLimiters[limit.name]
-          if (rateLimiter) {
-            this.routeRateLimiters[nsid].push({
-              consume: (ctx: XRPCReqContext) =>
-                rateLimiter.consume(ctx, {
-                  calcKey,
-                  calcPoints,
-                }),
-              reset: (ctx: XRPCReqContext) =>
-                rateLimiter.reset(ctx, {
-                  calcKey,
-                }),
-            })
+  private createRouteRateLimiter(
+    nsid: string,
+    config: XRPCHandlerConfig,
+  ): RouteRateLimiter | undefined {
+    const { rateLimits } = this.options
+
+    // @NOTE Silently ignore creation of route rate limiter if the `rateLimits`
+    // options was not provided to the constructor.
+    if (rateLimits && config.rateLimit) {
+      const rateLimiters = asArray(config.rateLimit).map((options, i) => {
+        if (isShared(options)) {
+          const rateLimiter = this.sharedRateLimiters.get(options.name)
+          if (!rateLimiter) {
+            // Config error
+            throw new TypeError(
+              `Shared rate limiter "${options.name}" not found`,
+            )
           }
+
+          return updateRateLimiterOptions(rateLimiter, options)
         } else {
-          const { durationMs, points } = limit
-          const rateLimiter = this.options.rateLimits?.creator({
-            keyPrefix: `nsid-${i}`,
-            durationMs,
-            points,
-            calcKey,
-            calcPoints,
-          })
-          if (rateLimiter) {
-            this.sharedRateLimiters[nsid] = rateLimiter
-            this.routeRateLimiters[nsid].push({
-              consume: (ctx: XRPCReqContext) =>
-                rateLimiter.consume(ctx, {
-                  calcKey,
-                  calcPoints,
-                }),
-              reset: (ctx: XRPCReqContext) =>
-                rateLimiter.reset(ctx, {
-                  calcKey,
-                }),
-            })
-          }
+          return rateLimits.creator({ ...options, keyPrefix: `${nsid}-${i}` })
```

Comment:
Fixes an issue where the same prefix (`"nsid-0"`, `"nsid-1"`, ...) was used for every method defining a custom `rateLimit` config.

https://github.com/bluesky-social/atproto/blob/d1e3e68dd9eb7bed13d9023bc0e4ce3c448eabf5/packages/xrpc-server/src/server.ts#L470